KAFKA-5454: Add a new Kafka Streams example IoT oriented

ppatierno opened the pull request:
Added a Kafka Streams example (IoT oriented) using "tumbling" window

ppatierno committed:
Added a Kafka Streams example (IoT oriented) using "tumbling" window

Fix checkstyle errors on indentation and number of instructions per line

Fix valueOf to parseInt for temperature values

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5354/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5339/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5355/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5340/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5357/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5342/
Test PASSed (JDK 8 and Scala 2.12).

ewencp commented:
/cc @guozhangwang @dguy @mjsax  @bbejeck @enothereska

enothereska commented:
cc @miguno too

enothereska commented:
Thanks LGTM.

bbejeck commented:
LGTM,  thanks for the example.

guozhangwang commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,109 @@
+ * sent to a topic named "max" if it exceeds 20
+ *
+ * Before running this example you must create the input topic and the output topic (e.g. via
+ * bin/kafka-topics.sh --create ...), and write some data to the input topic (e.g. via
```
We need to specify what are the input message format needed for the value string: it seems needed to be an integer.

Also stating the full string of the cmd line tool would help.

guozhangwang commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,109 @@
+
+        // usually the stream application would be running forever,
+        // in this example we just let it run for some time (big enough for covering the used window) and then stop
+        Thread.sleep(30000);
```
How about changing this to be only stoppable by ctrl-C? We are changing the rest of the examples as well in a manner to improve our quick start: https://github.com/apache/kafka/pull/3515

guozhangwang commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,109 @@
+                            return value2;
+                    }
+                }, TimeWindows.of(5000))
+                .toStream()
```
Why we need to call `toStream()` before calling `filter` here?

ppatierno replied:
@guozhangwang what I see without using filter on a KStream but on a KTable instead is that : if in the 5 seconds window the ingested values are above 20 (so the filter result is FALSE), at the end of the window a "null" value is written into the output topic. It doesn't happen using a filter on a KStream. Is it an expected behaviour or a bug ?

guozhangwang replied:
`values are above 20` you mean `are below 20` right?

guozhangwang replied:
Just to clarify: I thought `filter` operator is needed for your logic, but `toStream` seems not needed. Since `KTable` itself also supports the `filter` operator so you do not need to force transforming it into a `KStream` before applying `filter`. But your observation is right, and that is by-design of the `KTable#filter` APIs:

https://issues.apache.org/jira/browse/KAFKA-3902

Let me know if you think this does not make sense.

ppatierno replied:
Yes I meant "are below 20" sorry. Regarding the filter on KTable I'm going to read the JIRA but before reading it I think that if I'm filtering I want to see only messages for which the predicate is TRUE flowing into the destination topic (in this case) and not having a message with "null" as body.

ppatierno replied:
@guozhangwang it seems that we are in the case c.3 from JIRA but does it mean something like "record deleted" in my understanding if the messages are useful for re-building a table. But is it our case ? I think that in this example a KStream seems to be more appropriate.

guozhangwang replied:
@ppatierno Okay, I think I finally get your points now. And that's the beauty of written communications :) What I had originally is that

1. Hmm, `toStreams` seems not needed since `KTable` has a `filter` operator as well.
2. Then I got a reply from you that seems to say `KTable#filter` seems incorrect that it prints unexpected outputs and that is why I have to use `KTable.toStream.filter()`.
3. Then I replied again saying well, that output is actually expected due to the `KTable#filter` semantics.

Now regarding to whether that semantics fit well with the app logic that depends on how we interpret the app. Let's say if your purpose is building an online monitoring system that alerts whenever degree is beyond threshold 20 on a windowing basis of 5 minutes, and then let's say there a degree readings of

```
10, 25, 15
```

For a specific window, the question is does it need to trigger an alert, since it "has ever exceeded the threshold", or not since "its final value within that degree is below the threshold"? For example:

1. In the former case, I think a windowing operation would be meaningless since we would have to trigger an alert (i.e. write that into the final output topic so that it can be read and alert triggered) whenever there is a value higher than threshold; but that seems to be your application logic, with `toStream.filter` your output messages are `25` only. But as I said you would not need a windowed reduce function at all but just `stream.filter().to()`.

2. In the latter case, then a windowing operation does make sense, but then you should not alert on `25` since that is not the final value within the window, and with `table.filter()` your output would be `null` indicating that for that window the final value does not actually exceed the threshold and hence we do not need to alert.

Does that make sense to you?

ppatierno replied:
@guozhangwang finally we are on the same page :-) My case was the 2 but ... I don't understand the need for having a `null` value in the destination alarm topic. I mean, the application which should raise the alarm is reading such topic and it should raise an alarm only if it receives a message. If the message arrives in any case even when there is no alarm with `null` body, the application does extra work : needs to read it, checks that it's `null` or not an then raises the alarm eventually.

guozhangwang replied:
With the input of `10, 25,15` and using `table.filter.to()` you would get two outputs: `25` and `null` indicating that it was once over 25 and then "not any more".

If the input is just, say, `15` ideally we should print nothing after `filter` but today admittedly we still print a `null`. So to walk around it I agree that maybe we can call `toStream` first. But the can be addressed when we improve the `filter` impl again.

ppatierno replied:
@guozhangwang so as far as understood this implementation is now good to you. Are you going to merge and close this PR ?

guozhangwang commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,109 @@
+        KStream<String, String> source = builder.stream("temperature");
+
+        KStream<Windowed<String>, String> max = source
+                .map(new KeyValueMapper<String, String, KeyValue<String, String>>() {
```
We can call `selectKey()` instead which is like a syntax-sugar for such cases.

ppatierno committed:
Added stopping example via "ctrl-C"

miguno commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Demonstrates, using the high-level KStream DSL, how to implement an IoT demo application
+ * which ingests temperature value processing the maximum value in the latest 5 seconds sending
+ * a new message if it exceeds 20
```
Nit: `.` full stop missing.

miguno commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,121 @@
+ *
+ * In this example, the input stream reads from a topic named "temperature", where the values of messages
+ * represent temperature values; using a 5 seconds "tumbling" window, the maximum value is processed and
+ * sent to a topic named "max" if it exceeds 20
```
Nit: `.` full stop missing.

miguno commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,121 @@
+ * represent temperature values; using a 5 seconds "tumbling" window, the maximum value is processed and
+ * sent to a topic named "max" if it exceeds 20
+ *
+ * Before running this example you must create the input topic and the output topic (e.g. via
```
I'd suggest to include the actual, full CLI commands to create the input topic ("temperature") and the output topic ("max").  We should make it as easy as possible for beginners to get started -- some user might end up playing with Kafka for the first time because they want to run this specific Kafka Streams example.

Also, I'd suggest to somehow prefix the topic names so that it is a bit easier to tie them back to this example?  Like: "temperature" -> "iot-temperature", "max", "iot-temperature-max"?

miguno commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,121 @@
+                .map(new KeyValueMapper<String, String, KeyValue<String, String>>() {
+                    @Override
+                    public KeyValue<String, String> apply(String key, String value) {
+                        return new KeyValue<>("temp", value);
```
We should explain why the key ("temp") is hard-coded here.

miguno commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/temperature/TemperatureDemo.java:
```diff
@@ -0,0 +1,121 @@
+                        else
+                            return value2;
+                    }
+                }, TimeWindows.of(5000))
```
I'd suggest to replace `5000` with `TimeUnit.SECONDS.toMillis(5)`.  This is better than magic numbers.

miguno reviewed (changes requested):
Thanks for the PR, @ppatierno ! :)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6295/
Test PASSed (JDK 7 and Scala 2.11).

ppatierno committed:
Changed the map into selectKey for assigning key to temperature messages
Renamed input and output topics
Defined constants for the temperature threashold and window size
Added CLI commands for creating topics and running consumer/producer

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6279/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6297/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6281/
Test FAILed (JDK 8 and Scala 2.12).

ppatierno committed:
Specified that input string values needs to be integers

ppatierno commented:
@miguno I have updated the PR with your change requests. I agree on them.

@guozhangwang the same for your suggestions. I didn't apply the only one related to using toStream() and the filter() (so on KStream instead of a KTable). See my reply.

Thanks guys !

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6299/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6283/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang commented:
@ppatierno Sorry for the late replies. I think the PR is good to go. Though there is a refactoring on the StreamsBuilder / Topology today which deprecates some of the public APIs and proposed a set of new APIs inside `o.a.k.streams` directly. Could you rebase this PR once again? Thanks!

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6452/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6437/
Test PASSed (JDK 8 and Scala 2.12).

ppatierno committed:
Updated example using the new StreamBuilder

ppatierno commented:
@guozhangwang I have rebased and updated the example using the new StreamBuilder. Now it should be ok ! Thanks ! ;)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6461/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6446/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang commented:
Merged to trunk, thanks!
